sretoolbox/binaries/base.py

```diff
@@ -0,0 +1,243 @@
+"""
+Abstractions around system binaries.
+"""
+
+import logging
+import os
+
+from abc import ABCMeta
+from abc import abstractmethod
+from collections import Counter
+from pathlib import Path
+
+import requests
+
+from semver import VersionInfo
+
+from sretoolbox.utils import run
+
+
+LOG = logging.getLogger(__name__)
+
+
+class CommandNotFoundError(Exception):
+    """
+    Used when a binary is not available in the system.
+    """
+
+
+class Binary(metaclass=ABCMeta):
+    """
+    Represents a binary in the system. Tries to find the system binary
+    on the $PATH, in the specified version. When the binary is not
+    available locally, tryes to download it.
+
+    :param version: the version of the binary
+    """
+    binary_template = ''
+    download_url_template = ''
+
+    def __init__(self, version, download_path='/tmp'):
```

Comment:
I don't like the default here too much. Since the binary is going to stay, I guess I'd prefer this being explicit so that users of this class are forced to properly handle the `download_path`

Reply:
I might not want it to stay... and it might even be already in the $PATH

Reply:
Sure. My point is that I don't like the idea of polluting the filesystem if class is instantiated without `download_path`. Nothing very critical, though :-D